e3nn/o3/_wigner.py

```diff
@@ -118,76 +130,139 @@ def wigner_3j(l1, l2, l3, flat_src=_W3j_flat, dtype=None, device=None):
-    def _DxDxD(a, b, c):
-        D1 = wigner_D(l1, a, b, c)
-        D2 = wigner_D(l2, a, b, c)
-        D3 = wigner_D(l3, a, b, c)
-        return torch.einsum('il,jm,kn->ijklmn', D1, D2, D3).reshape(n, n)
-
-    random_angles = torch.tensor([
-        [4.41301023, 5.56684102, 4.59384642],
-        [4.93325116, 6.12697327, 4.14574096],
-        [0.53878964, 4.09050444, 5.36539036],
-        [2.16017393, 3.48835314, 5.55174441],
-        [2.52385107, 0.29089583, 3.90040975],
-    ], dtype=dtype, device=device)
-
-    B = random_angles.new_zeros((n, n))
-    for abc in random_angles:
-        D = _DxDxD(*abc) - torch.eye(n, dtype=dtype, device=device)
-        B += D.T @ D
-
-    eigenvalues, eigenvectors = torch.linalg.eigh(B)
-    assert eigenvalues[0] < 1e-10
-    Q = eigenvectors[:, 0]
-    assert (B @ Q).norm() < 1e-10
-    Q = Q.reshape(2 * l1 + 1, 2 * l2 + 1, 2 * l3 + 1)
-
-    Q[Q.abs() < 1e-14] = 0
-
-    if Q[l1, l2, l3] != 0:
-        if Q[l1, l2, l3] < 0:
-            Q.neg_()
-    else:
-        if next(x for x in Q.flatten() if x != 0) < 0:
-            Q.neg_()
-
-    abc = o3.rand_angles(100, dtype=dtype, device=device)
-    Q2 = torch.einsum("zil,zjm,zkn,lmn->zijk", wigner_D(l1, *abc), wigner_D(l2, *abc), wigner_D(l3, *abc), Q)
-    assert (Q - Q2).norm() < 1e-10
-    assert abs(Q.norm() - 1) < 1e-10
-
-    return Q
+    assert isinstance(j1, (int, float))
+    assert isinstance(j2, (int, float))
+    assert isinstance(j3, (int, float))
+    mat = torch.zeros((int(2 * j1 + 1), int(2 * j2 + 1), int(2 * j3 + 1)), dtype=torch.float64)
+    if int(2 * j3) in range(int(2 * abs(j1 - j2)), int(2 * (j1 + j2)) + 1, 2):
+        for m1 in (x / 2 for x in range(-int(2 * j1), int(2 * j1) + 1, 2)):
+            for m2 in (x / 2 for x in range(-int(2 * j2), int(2 * j2) + 1, 2)):
+                if abs(m1 + m2) <= j3:
+                    mat[int(j1 + m1), int(j2 + m2), int(j3 + m1 + m2)] = su2_clebsch_gordan_coeff((j1, m1), (j2, m2), (j3, m1 + m2))
+    return mat
+
+
+def su2_clebsch_gordan_coeff(idx1, idx2, idx3):
+    """Calculates the Clebsch-Gordon coefficient
+    for SU(2) coupling (j1,m1) and (j2,m2) to give (j3,m3).
+    Parameters
+    ----------
+    j1 : float
+        Total angular momentum 1.
+    j2 : float
+        Total angular momentum 2.
+    j3 : float
+        Total angular momentum 3.
+    m1 : float
+        z-component of angular momentum 1.
+    m2 : float
+        z-component of angular momentum 2.
+    m3 : float
+        z-component of angular momentum 3.
+    Returns
+    -------
+    cg_coeff : float
+        Requested Clebsch-Gordan coefficient.
+    """
+    from fractions import Fraction
+    from math import factorial
+
+    j1, m1 = idx1
+    j2, m2 = idx2
+    j3, m3 = idx3
+
+    if m3 != m1 + m2:
+        return 0
+    vmin = int(max([-j1 + j2 + m3, -j1 + m1, 0]))
+    vmax = int(min([j2 + j3 + m1, j3 - j1 + j2, j3 + m3]))
+
+    def f(n):
+        assert n == round(n)
+        return factorial(round(n))
```

Comment:
I've found it to be a huge speedup to cache these factorials, in my code I use something like:

```
@functools.lru_cache(maxsize=None)
def factorial(n: int) -> int:
    return math.factorial(n)
```

`maxsize` might need to be set if you think there's a chance users could blow through their RAM.

Reply:
I am very surprised that factorial are slow enough to require cache!! 
I will give it a try

Reply:
Yes, it might be less impactful on typical e3nn workflows.  For my case, I was doing many thousands (or more) of transformations between geometric and spherical tensors so the factorials were a non-trivial component of my runtime.

Reply:
Maybe I'm wrong:

```
In [9]: %timeit math.factorial(1024)
48.6 µs ± 67.9 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Maybe my speed-up came from caching binomial coefficients that I also have to calculate...